From 0cf00853c9949689131c61580c82b0d4a9cdc935 Mon Sep 17 00:00:00 2001
Message-Id: <0cf00853c9949689131c61580c82b0d4a9cdc935.1425872624.git.chang-joon.lee@intel.com>
From: N Tajunnisha <tajunnisha.n@intel.com>
Date: Sat, 28 Feb 2015 00:29:04 +0530
Subject: [PATCH] REVERTME [VPG]: drm/i915: DSI port initialization patch

For single link MIPI DSI panels on port C, port value for
OTP, enable and disable sequence (from VBT block 53) is set
to 0 and it doesnot match the MIPI DSI port to enable.
Due to this, MIPI on port C is blanking out.

It is expected that for single link MIPI DSI panels, port
value from VBT Block 2 should be used.

In this patch, if the MIPI DSI panel is single link, then
fix is made to use port value from Block 2 to decide on
which port the OTP, enable and disable packets has to be sent
instead of using port id field from OTP, enable and disable
sequence (from vbt Block 53).

This is tagged as revertme as a fix for this issue exists in upstream.

Upstream-Repo: http://anongit.freedesktop.org/git/drm-intel
Upstream-commit: f915084edc5aad351d6a4675b0bcdded9a00090d
commit f915084edc5aad351d6a4675b0bcdded9a00090d
Author: Gaurav K Singh <gaurav.k.singh@intel.com>
Date:   Wed Dec 10 22:07:40 2014 +0530

    drm/i915: Changes related to the sequence port no for

    From now on for both DSI Ports A & C, the seq_port value has been
    set to 0. seq_port value is parsed from Sequence block#53 of VBT.
    So, for packets that needs to be read/write for DSI single link on
    Port A and Port C will now be based on the DVO port from VBT block 2
    instead of seq_port.

    Signed-off-by: Gaurav K Singh <gaurav.k.singh@intel.com>
    Reviewed-by: Jani Nikula <jani.nikula@intel.com>
    Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>

Issue: IMINAN-19711
Change-Id: If10027adf23316b46939892e2c5813f11be3b77d
Signed-off-by: Tajunnisha N <tajunnisha.n@intel.com>
Signed-off-by: Prathik Chandrashekar <prathik.chandrashekar@intel.com>
---
 drivers/gpu/drm/i915/intel_dsi_panel_vbt.c |   21 +++++++++++++++++++--
 1 file changed, 19 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c b/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c
index 1dd8bee..8b2f515 100644
--- a/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c
+++ b/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c
@@ -241,13 +241,30 @@ static struct gpio_table gtable[] = {
 
 static u8 *mipi_exec_send_packet(struct intel_dsi *intel_dsi, u8 *data)
 {
-	u8 type, byte, mode, vc, port;
+	struct drm_device *dev = intel_dsi->base.base.dev;
+	struct drm_i915_private *dev_priv = dev->dev_private;
+	u8 type, byte, mode, vc, port = 0;
 	u16 len;
 
 	byte = *data++;
 	mode = (byte >> MIPI_TRANSFER_MODE_SHIFT) & 0x1;
 	vc = (byte >> MIPI_VIRTUAL_CHANNEL_SHIFT) & 0x3;
-	port = (byte >> MIPI_PORT_SHIFT) & 0x3;
+
+	if (intel_dsi->dual_link)
+		port = (byte >> MIPI_PORT_SHIFT) & 0x3;
+	else {
+		/*
+		 * For single link, use port id from vbt block 2
+		 * instead of block 53
+		 */
+		if (dev_priv->vbt.dsi.port == DVO_PORT_MIPIA)
+			port = 0;
+		else if (dev_priv->vbt.dsi.port == DVO_PORT_MIPIC)
+			port = 1;
+		else
+			DRM_ERROR("Invalid port %d from VBT\n",
+					dev_priv->vbt.dsi.port);
+	}
 
 	/* LP or HS mode */
 	intel_dsi->hs = mode;
-- 
1.7.9.5

